[DM-16775] Implemented notebook to exercise SQuaSH InfluxDB data model

afausti opened the pull request:
You can use this notebook to try different strategies mapping `lsst.verify` data to InfluxDB. You can also use it to "manually" synchronize the SQuaSH production database with an InfluxDB instance.

afausti committed:
Added notebook to exercise data model

afausti committed:
Fixed flake8 E117 over-indented

SimonKrughoff commented on examples/influxdb_data_model.ipynb:
```diff
@@ -0,0 +1,511 @@
+   "source": [
+    "# SQuaSH InfluxDB data model\n",
+    "\n",
+    "In [DM-18103](https://jira.lsstcorp.org/browse/DM-18103) we revisit the SQuaSH InfluxDB data model. You can use this notebook to try different strategies mapping `lsst.verify` data to InfluxDB. You can also use it to \"manually\" synchronize the SQuaSH production database with an InfluxDB instance. For a quick introduction on InfluxDB concepts see [this notebook](https://github.com/lsst-sqre ...
```
Maybe point to the actual notebook instead of the repository.

SimonKrughoff commented on examples/influxdb_data_model.ipynb:
```diff
@@ -0,0 +1,511 @@
+   "metadata": {},
+   "source": [
+    "## Mapping SQuaSH metadata to InfluxDB\n",
+    "`lsst.verify` adds metadata to verification jobs and this is uploaded to SQuaSH. The mapping this metadata to either InfluxDB tags or fields is defined by this mapping:\n",
```
The mapping this --> The mapping of this

SimonKrughoff commented on examples/influxdb_data_model.ipynb:
```diff
@@ -0,0 +1,511 @@
+   "metadata": {},
+   "source": [
+    "The rationale for this mapping is the following:\n",
+    "1. `ci_id` and `squash_id` are sequential numbers if they are mapped to InfluxDB tag it will [increase InfluxDB series cardinality](https://docs.influxdata.com/influxdb/v1.7/concepts/schema_and_data_layout/#discouraged-schema-design) (DM-18342)\n",
```
tag --> tags

SimonKrughoff commented on examples/influxdb_data_model.ipynb:
```diff
@@ -0,0 +1,511 @@
+    "The rationale for this mapping is the following:\n",
+    "1. `ci_id` and `squash_id` are sequential numbers if they are mapped to InfluxDB tag it will [increase InfluxDB series cardinality](https://docs.influxdata.com/influxdb/v1.7/concepts/schema_and_data_layout/#discouraged-schema-design) (DM-18342)\n",
+    "2. It is not possible to do [math operations with InfluxDB timestamps](https://community.influxdata.com/t/math-operations-on-field-value-and-time/6323/4) so it is useful to add the `timestamp` explicitly as a field. There are also a `date` field which is added as environment metadata that we don't need in InfluxDB (DM-17049)\n",
+    "3. `ci_url` will be used to connected Chronograf to CI. It is a different URL for each CI run. The rationale for adding `ci_url` as an InfluxDB field it is the same as in 1 (DM-18342)\n",
```
I don't completely understand.  From point 3. above, it sounds like if `schema` is `None` the data will not be transferred to `influxDB`.  However, this statement makes it sound like `ci_url` is being put in a field.  Can you clarify what is happening?

afausti replied:
Fixed the mapping and improved the description, the `ci_url` is now correctly added as an InfluxDB field.

SimonKrughoff commented on examples/influxdb_data_model.ipynb:
```diff
@@ -0,0 +1,511 @@
+    "2. It is not possible to do [math operations with InfluxDB timestamps](https://community.influxdata.com/t/math-operations-on-field-value-and-time/6323/4) so it is useful to add the `timestamp` explicitly as a field. There are also a `date` field which is added as environment metadata that we don't need in InfluxDB (DM-17049)\n",
+    "3. `ci_url` will be used to connected Chronograf to CI. It is a different URL for each CI run. The rationale for adding `ci_url` as an InfluxDB field it is the same as in 1 (DM-18342)\n",
+    "4. We skip `packages` metadata for now, we plan to add to InfluxDB only the packages that changed between two consecutive CI runs (DM-18343)\n",
+    "5. `lsst.verify` metadata uses `filter_name` we decided to rename it to `filter` is the dataID key commonly used in DM.\n",
```
I think this should read:
`lsst.verify` metadata uses `filter_name`.  We decided to rename it since it is commonly called `filter` in the dataID used in processing with DM software.

afausti committed:
Handle metadata for the ldf environment

- Added ldf metatada in the mapping and handle
other edge cases in the code
- String fields are correctly added now
- Improved notebook narrative

afausti committed:
Describe mapping of lsst.verify concepts

SimonKrughoff commented on examples/influxdb_data_model.ipynb:
```diff
@@ -0,0 +1,511 @@
+    "            if new_key and schema == 'tag':\n",
+    "                tags.append(\"{}={}\".format(sanitize(new_key), sanitize(value)))\n",
+    "            elif new_key and schema == 'field':\n",
+    "                fields.append(\"{}={}\".format(sanitize(new_key), value))\n",
```
OK, this makes it clear that if the schema doesn't have a value of `tag` or `field`, it is dropped on the floor.  I think the description above about `ci_url` should be clarified.

SimonKrughoff commented on examples/influxdb_data_model.ipynb:
```diff
@@ -0,0 +1,511 @@
+    "    data = requests.get(SQUASH_API_URL + \"/job/{}\".format(job_id)).json()\n",
+    "    \n",
+    "    # Skip deprecated datasets\n",
+    "    if data['ci_dataset'] == 'unknown' or data['ci_dataset'] == 'decam':\n",
```
They've been telling us we can turn on decam in CI again.  I don't know if that changes how you want to handle things here.

afausti replied:
I would just remove this restriction once `decam` is turned on in CI again.

SimonKrughoff commented on examples/influxdb_data_model.ipynb:
```diff
@@ -0,0 +1,511 @@
+    "    \n",
+    "    fields_by_measurement = {}\n",
+    "    for verify_measurement in data['measurements']:\n",
+    "        # DM-XXXX - SQuaSH API /measurements should return the verification package \n",
```
Is there a ticket for this annotation?

SimonKrughoff reviewed (approved):
Looks good.  I have some comments that look to be on mostly outdated commits, but you may have a look anyway.

afausti committed:
Add a common data model for ci and ldf

- Execution environment metadata from ci and ldf have different
keys as stored in SQuaSH but we can have a common data model to
accomodate both.

afausti committed:
Addressed review comments
